velox/functions/prestosql/types/JsonType.cpp

```diff
@@ -62,57 +66,415 @@ void generateJsonTyped(
+// rows.
+template <
+    TypeKind kind,
+    bool throwIfNull = false,
+    typename std::enable_if<TypeTraits<kind>::isPrimitiveType, int>::type = 0>
+void castToJson(
     const BaseVector& input,
+    exec::EvalCtx* /*context*/,
     const SelectivityVector& rows,
     FlatVector<StringView>& flatResult) {
   using T = typename TypeTraits<kind>::NativeType;
 
+  // input is guaranteed to be in flat or constant encodings when passed in.
   auto inputVector = input.as<SimpleVector<T>>();
 
   std::string result;
   rows.applyToSelected([&](auto row) {
-    result.clear();
-    generateJsonTyped(*inputVector, row, result);
+    if (inputVector->isNullAt(row)) {
+      if constexpr (throwIfNull) {
+        VELOX_FAIL("Unexpected null value in input vector.");
+      } else {
+        flatResult.set(row, "null");
+      }
+    } else {
+      result.clear();
+      generateJsonTyped(*inputVector, row, result);
 
-    flatResult.set(row, StringView{result});
+      flatResult.set(row, StringView{result});
+    }
   });
 }
 
+// Forward declaration.
+void castToJsonFromArray(
+    const BaseVector& input,
+    exec::EvalCtx* context,
+    const SelectivityVector& rows,
+    FlatVector<StringView>& flatResult);
+
+void castToJsonFromMap(
+    const BaseVector& input,
+    exec::EvalCtx* context,
+    const SelectivityVector& rows,
+    FlatVector<StringView>& flatResult);
+
+void castToJsonFromRow(
+    const BaseVector& input,
+    exec::EvalCtx* context,
+    const SelectivityVector& rows,
+    FlatVector<StringView>& flatResult);
+
+// Casts complex-type input vectors to Json type.
+template <
+    TypeKind kind,
+    typename std::enable_if<!TypeTraits<kind>::isPrimitiveType, int>::type = 0>
+void castToJson(
+    const BaseVector& input,
+    exec::EvalCtx* context,
+    const SelectivityVector& rows,
+    FlatVector<StringView>& flatResult) {
+  if constexpr (kind == TypeKind::ARRAY) {
+    castToJsonFromArray(input, context, rows, flatResult);
+  } else if constexpr (kind == TypeKind::MAP) {
+    castToJsonFromMap(input, context, rows, flatResult);
+  } else if constexpr (kind == TypeKind::ROW) {
+    castToJsonFromRow(input, context, rows, flatResult);
+  } else {
+    VELOX_NYI(
+        "Casting from {} to JSON is not supported yet.",
+        input.type()->toString());
+  }
+}
+
+// Estimates the total length of all Json strings for a vector of arrays or maps
+// from the Json strings of their elements. The elementsStrings vector is
+// wrapped in elementsIndices. This estimate is an upperbound of the actual
+// total length.
+FOLLY_ALWAYS_INLINE size_t estimateJsonStringSizeForArraysAndMaps(
+    const FlatVector<StringView>& elementsStrings,
+    const vector_size_t* elementsIndices,
+    const vector_size_t* topLevelOffsets,
+    const vector_size_t* topLevelSizes,
+    const SelectivityVector& topLevelRows) {
+  size_t sum = 0;
+
+  topLevelRows.applyToSelected([&](auto row) {
```

Comment:
what if the array at row is null?

Reply:
You are right. I initially thought null rows should have been de-selected. But when arrays are elements of another complex-type vector, there could be null rows. I've corrected it in the new version. Thank you very much for pointing this out!